Move toward sharing markdown code with mobile.

showell opened the pull request:
The test plan here is fairly obvious--send some messages from the client and make sure their local-echo versions look ok.  We also have good automated test coverage here.

This PR makes significant progress toward resolving #13892.

timabbott commented:
For the `util.js` stuff, why not move those into `shared/util.js`?  Well, we'd need another name, but it seems like that'd be a better way to handle `escape_html` and friends than duplicating.

andersk commented:
I don’t think most of `util.js` belongs in `shared` (especially given its dependency on `blueslip`, but even ignoring that). We have `_.escape` for HTML escaping.

timabbott commented:
Yeah, I guess using `_.escape` here is a reasonable option.  But note that `markdown` did previous use a bunch more pieces of `util` around the subject->topic transition; I'm had thought we wanted that backwards-compatibility code to be available for mobile use.  But if we don't, it's simpler to just cut the dependency on util.

andersk commented:
I’ve dusted off #13524, which removes `escape_html` in favor of `_.escape` while correcting other bugs.

showell commented:
I re-pushed to address the comments from my first push:

* We're holding off on putting util.js into shared (instead, I am only removing it from `window.` for now)
* I've rebased with the escape-related fixes from @andersk 
* I've split out separate commits for the subject/topic related helpers and tried to make the commit messages provide a little more context

I've also added stuff since the first push, so there's more here to review.

gnprice commented:
Thanks @showell!

Some comments on the underscore / `escape` question:

    Because we are about to share the markdown library
    with mobile, there will be two possible variations
    of escape.  Mobile currently uses lodash; the
    webapp uses underscore.js.  Also, it's not clear
    if mobile has `_` in the global namespace.

Indeed, we don't put `_` in the global namespace. OTOH the shared package already depends on it (as specified in `static/shared/package.json`); the way it uses it is through this line in `typing_status.js`:
```js
import _ from "underscore";
```
So it'd be harmless to add more uses of underscore; it should just be explicitly imported in any module that uses it.

We do use `lodash.escape` for escaping input as HTML in the HTML we generate in the mobile app's own codebase, in the templates for things like recipient headers. So we could pass that in with no trouble.

But I'd also be quite content, and it seems like it might be simpler within the web and shared code, for the Markdown implementation in the shared code to just rely on the shared code's existing dependency on underscore and call `_.escape` directly. (After importing it explicitly at the top of the file.) Then whenever the webapp's migration to lodash reaches `_.escape`, the shared code can be included in that sweep; it'll just need to gain a `lodash.escape` dependency in its `package.json`.

andersk commented on static/js/markdown_config.js:
```diff
@@ -0,0 +1,54 @@
+        get_user_by_name: people.get_by_name,
+        get_user_by_id: people.get_by_user_id,
+        is_known_user_id: people.is_known_user_id,
+        is_my_user_id: people.is_my_user_id,
```
This can be replaced with `=== my_user_id()`.

showell replied:
The function `is_my_user_id` has a long sordid history dating back to when we were fixing all the places the codebase was using `email` as a key.  More recently, it was defending against type confusion between strings and ints.  I think we're ready to kill this off completely, as it's had a `blueslip.error` clause for a couple weeks now, and I'm not aware of any errors being triggered.

I'm not quite ready to sweep that for the entire codebase (and I'm not implying you were suggesting that, to be clear, I'm just digressing), but I will definitely fix it in the markdown code.

Thanks!

andersk commented on static/js/markdown_config.js:
```diff
@@ -0,0 +1,54 @@
+        // user stuff
+        get_user_by_name: people.get_by_name,
+        get_user_by_id: people.get_by_user_id,
+        is_known_user_id: people.is_known_user_id,
```
This can be replaced with `get_by_user_id(…) !== undefined`.

showell replied:
The original motivation for `is_known_user_id` here is that we could avoid spammy blueslip errors from the subsequent call to `get_by_user_id`.  Unlike most places in the code, trying to get a person based on a user_id from our duplicate-names-are-possible-so-put-ids-in-the-mention syntax and then failing to get a user doesn't mean that we've broken some invariant in the code; i.e. it's just bad data from users (who may have editing the mention after autocomplete).

We now have an `ignore_missing` parameter in `get_by_user_id`, though, so there's really no need for `known_user_id`.

Here's the relevant code for context:

~~~ js
103         userMentionHandler: function (name, silently) {
104             let person = helpers.get_user_by_name(name);
105
106             const id_regex = /(.+)\|(\d+)$/g; // For @**user|id** syntax
107             const match = id_regex.exec(name);
108             if (match) {
109                 const user_id = parseInt(match[2], 10);
110                 if (helpers.is_known_user_id(user_id)) {
111                     person = helpers.get_user_by_id(user_id);
112                     if (person.full_name !== match[1]) { // Invalid Syntax
113                         return;
114                     }
115                 }
116             }
~~~

It turns out this code also has issues that @gnprice brought up in chat, though, which is that we shouldn't be asking mobile for "person" objects.  Instead, we should just have something `maybe_get_full_name_for_user_id`.

Anyway, I will address this.

showell replied:
Also, that entire function is a mess. Ugh.

gnprice commented:
One aspect of the interface, where mobile's needs differ from the webapp's because of differences in the architecture of how the two apps manage their data:

To fully make the interface something that fits in smoothly on both sides, we'll want the exposed functions to be pure functions of their arguments -- given the same arguments, they're return the same values every time. This is important in the mobile app because of our use of [Redux](https://redux.js.org/introduction/three-principles) to structure how we maintain our data structures.

Here, this means that `helpers` should be taken as a parameter by the functions in this module each time they're called, rather than set once. In the case of the webapp, it'll probably want to just pass the same `helpers` object each time; but in the mobile app, we'll instead pass each time a helpers object where all the getters in it close over the current, immutable, global state tree.

This also means we'll want to move `realm_filter_map` and `realm_filter_list` to live outside the module, and pass in getters for them like for the other application data it needs.

Another thing Steve mentioned in chat discussion just now is that it'd be good to move to getters like `get_full_name_for_userid`, returning primitive values, rather than getters like `get_user_by_id`. That will be helpful for us on mobile because our object types for users, etc., aren't quite the same as in the webapp.

(I think all of the above probably makes sense as additional commits following the ones already in this PR.)

showell commented on static/js/markdown_config.js:
```diff
@@ -0,0 +1,54 @@
+*/
+
+exports.get_helpers = () => {
+    return {
```
Fixed (haven't pushed yet).

showell commented on static/js/markdown_config.js:
```diff
@@ -0,0 +1,54 @@
+        translate_emoticons_to_names: emoji.translate_emoticons_to_names,
+
+        // realm filters
+        get_realm_filters: () => page_params.realm_filters,
```
I'm sure what you're proposing as the alternative here.  Do you want me to pass in data instead of a function?

showell replied:
Ok, I looked at this more closely.  Most of the confusion here is the name of `set_realm_filters`, which should instead be `build_realm_filter_rules` or something similar.  Will address.

showell commented on static/js/markdown_config.js:
```diff
@@ -0,0 +1,54 @@
+        get_stream_by_name: stream_data.get_sub,
+
+        // emojis
+        get_active_realm_emojis: () => emoji.active_realm_emojis,
```
I extracted get_emoji_url() here (which is parallel to get_emoji_name and get_emoji_codepoint).

showell commented on static/js/markdown_config.js:
```diff
@@ -0,0 +1,54 @@
+        get_active_realm_emojis: () => emoji.active_realm_emojis,
+        get_emoji_name: emoji.get_emoji_name,
+        get_emoji_codepoint: emoji.get_emoji_codepoint,
+        translate_emoticons_to_names: emoji.translate_emoticons_to_names,
```
That function relies on `emoticon_conversions`, which we get from `emoji_codes.json`, which is already consumed and munged by `emoji.js`.  We could have markdown read in that file directly, but I don't think we want to hardwire markdown to that implementation detail.

Can we leave this in for now and then maybe remove it later as part of an effort to share the emoji infrastructure with mobile?

showell replied:
Oh gosh.  I just looked more deeply at this function, and it's re-building a new list of emoticon regexes every time we apply markdown to a message.  That's pretty gross.  And you are correct that most of this code is pretty markdown-specific.  I'll try to fix this in a prep commit.

showell replied:
Fixed in https://github.com/zulip/zulip/commit/5e9df7a0a60949325d1e74b998325d8940a6a861

showell commented:
Tim, I believe it's safe to merge everything here, but particularly the commits before the big one (" markdown: Add helper configuration for mobile.").

Most of the feedback here isn't really a blocker for the current PR.  There are some places where my changes just expose warts in the current markdown implementation.  And there are some things we'll want to do before actually sharing our code with mobile, but I view most things here as forward progress (i.e. even if the structure isn't yet perfect for mobile, it's at least more malleable than before).

timabbott commented on static/js/emoji.js:
```diff
@@ -18,6 +18,20 @@ const zulip_emoji = {
     deactivated: false,
 };
 
+exports.get_emoji_url = (emoji_name) => {
```
This function name is no good -- this function __only__ looks are realm emoji, not unicode emoji, which is confusing and given the name, and likely not something we want to export/

timabbott replied:
Maybe what we want to do is extract a different function that returns the `emoji_data` object, and then just have the `markdown.js` caller access the data attributes of the object, and remove the direct `markdown.js` access to `emoji_codes`?  The interface problem here may be how tightly the `markdown` code digs into the emoji data structures.

timabbott commented:
I merged about half of this and then pushed back here.  I'm really excited about this!

showell commented:
Re-pushed.  The first three commits should be easy to push (all just phasing out or killing off util-related stuff).

timabbott commented:
Great, merged those first 3 commits.

showell commented:
Greg said:
> Another thing Steve mentioned in that chat discussion just now is that it'd be good to move to getters like `get_full_name_for_userid`, returning primitive values, rather than getters like `get_user_by_id`. That will be helpful for us on mobile because our object types for users, etc., aren't quite the same as in the webapp.

In the current PR we only have two have functions that deal with non-primitive types:
* get_user_group_from_name - This would be fairly easy to clean up; I just haven't gotten to it yet.
* get_stream_by_name - This is also fairly easy to clean up, but it can possibly wait for us to simply add `hash_util` to the shared library (which isn't quite trivial, but it's a really good thing to share for consistency-between-apps reasons).  The only place we use get_stream_by_name is for building hashses.

timabbott commented:
I renamed `get_emoji_url` to `get_realm_emoji_url` for clarity, updated some comments in that commit, and merged the 4 prep commits.  Still reading the last few.

timabbott commented:
I merged the mention commit as well.

timabbott commented:
And merged the last bit too!  The helper commit ended up pretty clean after all the refactoring.

I imagine we may do some more fiddling with the interface, but I think we can readily do that on top of this.  Huge thanks for making this happen @showell!  

I guess the next steps are to make it a proper module and then move the file to `static/shared/js/markdown.js`.
